### PR TITLE
Updating seqweaver

### PR DESCRIPTION
**Pull request by aviyalitman**
#### What does this implement/fix? Explain your changes.
One new script (selene_sdk/updating_seqweaver.py) implements a class which constructs and trains new data for seqweaver.
selene_sdk/targets/genomic_features.py was modified to add strand specificity when parsing the bed file. I had to change the indexing order in order to maintain the following bed file column order: [chr, start, end, strand, target]. 
selene_sdk/targets/_genomic_features.pyx was modified to fix the indexing after adding the strand column.

#### What testing did you do to verify the changes in this PR?
I tested the update_seqweaver.py script, which uses the other scripts as an API, using the original training dataset for Seqweaver (~218 CLIP-seq datasets for 90 RBPs).

### CODE REVIEW

File: selene_sdk/targets/_genomic_features.pyx [hunk 1]
```diff
@@ -31,7 +31,7 @@ def _fast_get_feature_data(int start,
         feature_end = int(row[2])
         index_start = max(0, feature_start - start)
         index_end = min(feature_end - start, query_length)
-        index_feat = feature_index_dict[row[3]]
+        index_feat = feature_index_dict[row[4]]
```
Comment (kathyxchen) [inline selene_sdk/targets/_genomic_features.pyx:34]: If we want to merge these into Selene in the future, one thing that will be important is backwards compatibility.

Right now, most users of Selene expect `row[3]` to be the target label that is being predicted. So for the addition of the strand column, I think we can do one or the other:
(1) Add the strand col as the fourth column for now (most minimal change, but maybe doesn't make as much sense from BED file column organization standpoint)
(2) Add the strand col as the third col and the targets col as the fourth column (what you have now) but our library needs to know to check---whether through specification of an optional parameter or automatically---the correct column for retrieving strand and targets now.

Just something worth thinking about! We can discuss what is the best way. 

You know your own priorities and schedule for this project best: if it will be too time-consuming to make the changes sufficient for pushing this to Selene within your timeline, you should focus on getting a model trained, and then if you have time, refactoring and testing your code in this pull request to make it ready for merging into Selene.

File: selene_sdk/targets/genomic_features.py [hunk 2]
```diff
@@ -133,7 +135,7 @@ def _get_feature_data(chrom, start, end,
         `i`th feature is positive, and zero otherwise.
 
     """
-    rows = get_feature_rows(chrom, start, end)
+    rows = get_feature_rows(chrom, start, end, strand)
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:138]: did you change this method definition somewhere so that it also optionally takes in `strand` as an argument? I might have missed it

File: selene_sdk/targets/genomic_features.py [hunk 3]
```diff
@@ -303,7 +305,7 @@ def dfunc(self, *args, **kwargs):
             return func(self, *args, **kwargs)
         return dfunc
 
-    def _query_tabix(self, chrom, start, end):
+    def _query_tabix(self, chrom, start, end, strand):
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:308]: make `strand` optional here too

File: selene_sdk/targets/genomic_features.py [hunk 4]
```diff
@@ -329,12 +334,13 @@ def _query_tabix(self, chrom, start, end):
 
         """
         try:
-            return self.data.query(chrom, start, end)
+            tabix_query = self.data.query(chrom, start, end)
+            return [line for line in tabix_query if str(line[3]) == strand] # strand specificity
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:338]: and then here add something like...
```
if strand == '+' or strand == '-':
  return [line for line in tabix_query if str(line[3]) == strand]
else:
  return tabix_query
```

File: selene_sdk/targets/genomic_features.py [hunk 5]
```diff
@@ -329,12 +334,13 @@ def _query_tabix(self, chrom, start, end):
...
         except tabix.TabixError:
             return None
 
     @init
-    def is_positive(self, chrom, start, end):
+    def is_positive(self, chrom, start, end, strand):
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:343]: make strand optional here too

File: selene_sdk/targets/genomic_features.py [hunk 6]
```diff
@@ -357,11 +363,11 @@ def is_positive(self, chrom, start, end):
             assume the error was the result of no features being present
             in the queried region and return `False`.
         """
-        rows = self._query_tabix(chrom, start, end)
+        rows = self._query_tabix(chrom, start, end, strand)
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:366]: `self._query_tabix(chrom, start, end, strand=strand)`

File: selene_sdk/targets/genomic_features.py [hunk 7]
```diff
@@ -388,14 +396,14 @@ def get_feature_data(self, chrom, start, end):
         """
         if self._feature_thresholds_vec is None:
             features = np.zeros(self.n_features)
-            rows = self._query_tabix(chrom, start, end)
+            rows = self._query_tabix(chrom, start, end, strand) # added strand specificity
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:399]: `self._query_tabix(chrom, start, end, strand=strand)`

File: selene_sdk/targets/genomic_features.py [hunk 8]
```diff
@@ -388,14 +396,14 @@ def get_feature_data(self, chrom, start, end):
...
             if not rows:
                 return features
             for r in rows:
-                feature = r[3]
+                feature = r[4]
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:403]: same comment on indexing here as my first very long comment

File: selene_sdk/update_seqweaver.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                sstart, ssend = self._from_midpoint(start, end)
+
+                # 1 x 4 x 1000 bp
+                # get_encoding_from_coords : Converts sequence to one-hot-encoding for each of the 4 bases
+                dna_seq, has_unk = seqs.get_encoding_from_coords_check_unk(chrom, sstart, ssend, strand=strand)
+                if has_unk: continue
+
+                # 1 x n_features
+                # get_feature_data: Computes which features overlap with the given region.
+                labels = targets.get_feature_data(chrom, start, end, strand)
```
Comment (kathyxchen) [inline selene_sdk/update_seqweaver.py:108]: `chrom, start, end, strand=strand`

**Review by kathyxchen — COMMENTED** [review]
Thank you @aviyalitman for this PR!

Everything looks good so far. I left a few minor comments about backwards compatibility, making `strand` an optional argument in general. Don't worry about making all these changes if you don't have a lot of time to work on the code - prioritize based on your schedule and we can always keep this PR around for future incorporation into Selene, if you want to focus on getting something working and models trained first. (:

I really appreciate all the comments and documentation you included with your changes, it makes your code very easy to review and understand!

**Comment by kathyxchen** [comment]
@aviyalitman let me know whenever this PR is ready for review - no rush!

**Comment by aviyalitman** [comment]
Hey @kathyxchen - it's ready for review!

I implemented the changes we talked about, specifically adding the strand column as the fourth column for now and making strand information completely optional.

I tested it on the seqweaver data. :)

File: models/seqweaver.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
+import torch
```
Comment (kathyxchen) [inline models/seqweaver.py:1]: add a little docstring before the imports saying this is the seqweaver architecture, cite which publication it's from (i might have forgotten to do this in the other `models` architecture files!)

File: models/seqweaver.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+                Lambda(lambda x: x.view(1, -1)
+                       if 1 == len(x.size()) else x),
+                nn.Linear(n_classes, n_classes)
+            ),
+            nn.Sigmoid(),
+        )
+
+    def forward(self, x):
+        x = x.unsqueeze(2)
+        #print('forward', x.shape)
```
Comment (kathyxchen) [inline models/seqweaver.py:56]: remove these commented out lines

File: selene_sdk/targets/genomic_features.py [hunk 12]
```diff
@@ -125,6 +125,8 @@ def _get_feature_data(chrom, start, end,
     get_feature_rows : types.FunctionType
         A function that takes coordinates and returns rows
         (`list(tuple(int, int, str))`).
+    strand : {'+', '-', '.'}, optional
+            Default is '+'. The strand the sequence is located on. '.' is treated
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:129]: what does `'.' is treated` mean?

File: selene_sdk/targets/genomic_features.py [hunk 13]
```diff
@@ -317,6 +319,9 @@ def _query_tabix(self, chrom, start, end):
             The 0-based start position of the query coordinates.
         end : int
             One past the last position of the query coordinates.
+        strand : {'+', '-', '.'}
+            The strand the sequence is located on. '.' is treated
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:323]: same comment for this line about `'.'` case

File: selene_sdk/update_seqweaver.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    def __init__(self, input_path, train_path, validate_path, feature_path, hg_fasta, yaml_path, VAL_PROP=0.01, sequence_len=1000):
+        """
+        Constructs a new `UpdateSeqweaver` object.
+        """
+        self.input_path = input_path
+        self.train_path = train_path
+        self.validate_path = validate_path
+        self.feature_path = feature_path
+        self.yaml_path = yaml_path
+        self.VAL_PROP = VAL_PROP
```
Comment (kathyxchen) [inline selene_sdk/update_seqweaver.py:52]: why is this in all caps, out of curiosity? is this some code convention you've seen before?

File: selene_sdk/update_seqweaver.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+from train_model import TrainModel
+from utils.config import load_path
+from utils.config_utils import parse_configs_and_run
+
+class UpdateSeqweaver():
+    """
+    Stores a dataset specifying sequence regions and features.
+    Accepts a tabix-indexed `*.bed` file with the following columns,
+    in order:
+        [chrom, start, end, strand, feature]
```
[the remark this answers is not in the document]
Comment (kathyxchen) [reply selene_sdk/update_seqweaver.py:24]: i do think this BED file format would be more desirable when including strand information (vs having `chrom, start, end, feature, strand`) but it'd be a more substantial change to GenomicFeatures, so maybe we can think about it as a future TODO...

File: selene_sdk/update_seqweaver.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
+"""
```
Comment (kathyxchen) [inline selene_sdk/update_seqweaver.py:1]: this is nicely written! I think you can put this in `selene_sdk/scripts` for the time being, but at some point I'd expect you would move this into a totally separate repository in `FunctionLab`, complete with downloadable files and a command-line script to be able to run this code and reproduce your work! 

it's a good reminder that maybe we should start compiling a list of projects that use selene for future users to be able to refer back to when they start their own projects

**Review by kathyxchen — CHANGES_REQUESTED** [review]
thanks - great work on this @aviyalitman! :) I had some minor comments / questions, let me know when you've gotten a chance to review those!

File: selene_sdk/targets/genomic_features.py (hunk 12 above)
Comment (aviyalitman) [reply selene_sdk/targets/genomic_features.py:129]: My bad - I wanted to be consistent with the rest of the code but this part got cut off :P it's fixed now.

File: selene_sdk/update_seqweaver.py (hunk 14 above)
Comment (aviyalitman) [reply selene_sdk/update_seqweaver.py:52]: Good point - I usually put global variables in all caps, which this variable temporarily was during testing but isn't anymore. Changed to lower caps :)

File: selene_sdk/update_seqweaver.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (aviyalitman) [reply selene_sdk/update_seqweaver.py:24]: I definitely agree that the previous order is more intuitive, this was somewhat of a quick fix. I thought about doing this without changing the column order but as you say, it will require substantial changes to GenomicFeatures. I would definitely like to go back to this soon :-)

File: selene_sdk/targets/genomic_features.py [hunk 17]
```diff
@@ -317,6 +319,9 @@ def _query_tabix(self, chrom, start, end):
             The 0-based start position of the query coordinates.
         end : int
             One past the last position of the query coordinates.
+        strand : {'+', '-', '.'}
+            The strand the sequence is located on. '.' is treated as '.'.
```
Comment (kathyxchen) [inline selene_sdk/targets/genomic_features.py:323]: sorry I do have one last thought on this - I think `'.' is treated as '.'` is still too ambiguous for someone who might not look at the code implementation explicitly (especially in other functions where you'd have to trace back to this specific function to see how it gets handled). 

You're basically saying that if the query is `chr, start, end, '.'` we wouldn't retrieve sequences that are `+` or `-`, only `.` right? I think the documentation should say that explicitly. 

That said.. I wonder if this should be handled differently. That is:

- if '+' or '-' is passed in, only retrieve rows with the correct matching strand -- exactly what you are doing now
- if '.' is passed in, treat it the same as strand = None (i.e. change your check to `if strand == '+' or strand == '-'` in line 338) 

does that sound reasonable or is there a situation I'm missing where it would be important to only retrieve rows with `.`?

**Review by kathyxchen — COMMENTED** [review]
LGTM minus the remaining thought about `strand='.'` handling - let me know your thoughts on that as I'm not sure if what I'm suggesting is right or not?

File: selene_sdk/targets/genomic_features.py (hunk 17 above)
Comment (aviyalitman) [reply selene_sdk/targets/genomic_features.py]: Fix pushed - I agree the handling of '.' was confusing before and think this approach is better! '.' is now treated as None.

**Review by kathyxchen — APPROVED** [review]
perfect, thank you @aviyalitman ! this PR is ready to merge (:
